src/graph/formulas/fallback.rs

```diff
@@ -0,0 +1,238 @@
+where
+    N: Node,
+    E: Edge,
+{
+    pub(crate) prefer_meters: bool,
+    pub(crate) graph: &'a ComponentGraph<N, E>,
+}
+
+impl<'a, N, E> FallbackExpr<'a, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    fn generate(&self, mut component_ids: BTreeSet<u64>) -> Result<Expr, Error> {
+        let mut formula = None::<Expr>;
+        while let Some(component_id) = component_ids.pop_first() {
+            if let Some(expr) = self.meter_fallback(component_id)? {
+                formula = Self::add_to_option(formula, expr);
+            } else if let Some(expr) = self.component_fallback(&mut component_ids, component_id)? {
+                formula = Self::add_to_option(formula, expr);
+            } else {
+                formula = Self::add_to_option(formula, Expr::component(component_id));
+            }
+        }
+
+        formula.ok_or(Error::internal("Search for fallback components failed."))
+    }
+
+    /// Returns a fallback expression for a meter component.
+    fn meter_fallback(&self, component_id: u64) -> Result<Option<Expr>, Error> {
+        let component = self.graph.component(component_id)?;
+        if !component.is_meter() || self.graph.has_meter_successors(component_id)? {
+            return Ok(None);
+        }
+
+        if !self.graph.has_successors(component_id)?
+            || !self
+                .graph
+                .successors(component_id)?
+                .all(|x| x.is_supported())
+        {
+            return Ok(Some(Expr::component(component_id)));
+        }
+
+        let successor_expr = self
+            .graph
+            .successors(component_id)?
+            .map(Expr::from)
+            .reduce(|a, b| a + b)
+            .ok_or(Error::internal(
+                "Can't find successors of components with successors.",
+            ))?;
+
+        let exprs = if self.prefer_meters {
+            vec![Expr::component(component_id), successor_expr]
+        } else {
+            vec![successor_expr, Expr::component(component_id)]
+        };
+
+        Ok(Some(Expr::coalesce(exprs)))
+    }
+
+    /// Returns a fallback expression for components with the following categories:
+    ///
+    /// - CHP
+    /// - Battery Inverter
+    /// - PV Inverter
+    /// - EV Charger
+    fn component_fallback(
+        &self,
+        component_ids: &mut BTreeSet<u64>,
+        component_id: u64,
+    ) -> Result<Option<Expr>, Error> {
+        let component = self.graph.component(component_id)?;
+        if !component.is_battery_inverter()
+            && !component.is_chp()
+            && !component.is_pv_inverter()
+            && !component.is_ev_charger()
+        {
+            return Ok(None);
+        }
+
+        let mut formula = None::<Expr>;
+
+        // If predecessors have other successors that are not in the list of
+        // component ids, the predecessors can't be used as fallback.
+        let siblings = self
+            .graph
+            .siblings_from_predecessors(component_id)?
```

Comment:
This and next commit are in wrong order. :)
Could you please switch them? Otherwise it is hard to read.